### ISSUE TITLE
avoid race condition when multiple tasks update the same file? 

### ISSUE THREAD

**Issue by biocyberman** [pydoit/doit#230]
This snippet is a function that update `current.manifest.json` file:

```python
current_dir = path.dirname(__file__)
current_manifest = path.join(current_dir, 'current.manifest.json')
target_manifest = path.join(current_dir, 'target.manifest.json')

def add_to_manifest(key):
    """Copy the JSON objects from the target JSON file to current JSON file"""
    with open(target_manifest, 'r') as target, \
            open(current_manifest, 'r+') as current:
        target_json = json.load(target)
        current_json = json.load(current)
        current_json[key] = target_json[key]
        current.seek(0)
        current.truncate()
        json.dump(current_json, current, indent=4)
```

when I run my tasks in parallel (`doit -n 4`) which call to this function in parallel, I got this error:

```python
TaskError - taskid:install_vcfanno
PythonAction Error
Traceback (most recent call last):
  File "/conda-env/lib/python3.6/site-packages/doit/action.py", line 403, in execute
    returned_value = self.py_callable(*self.args, **kwargs)
  File "/tasks/install/install_tools.py", line 170, in action
    add_to_manifest('vcfanno')
  File "/tasks/nectar/nectar_util.py", line 24, in add_to_manifest
    current_json = json.load(current)
  File "/conda-env/lib/python3.6/json/__init__.py", line 299, in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "/conda-env/lib/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "s/conda-env/lib/python3.6/json/decoder.py", line 339, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/conda-env/lib/python3.6/json/decoder.py", line 357, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

The problem here is that ` current.truncate()` empty the `current.manifest.json` when another task tries to load it at `current_json = json.load(current)`. 

I am thinking about using [file lock](https://pypi.python.org/pypi/filelock) to avoid this problem, but just want to check if there is a elegant solution already.

**Comment by biocyberman** [comment]
Since `doit` uses `multiprocessing`, I came up with this solution:


```python
def safe_add_to_manifest(key):
    """Copy the JSON objects from the target JSON file to current JSON file"""
    lock.acquire()  # Safety lock <--
    with open(target_manifest, 'r') as target, \
            open(current_manifest, 'r+') as current:

        target_json = json.load(target)
        current_json = json.load(current)
        current_json[key] = target_json[key]
        current.seek(0)
        current.truncate()
        json.dump(current_json, current, indent=4)

    lock.release() # <-- Release safety lock

```

And the lock is instantiated in `dodo.py`:

```python
....
from multiprocessing import Lock
lock = Lock()
....

**Comment by schettino72** [comment]
Yes, doit can not do much about this. This is logic specific to your application.
